StortSpelProjekt/Engine/src/Renderer/TextManager.cpp

```diff
@@ -0,0 +1,274 @@
+#include <vector>
+#include "TextManager.h"
+
+#include "../Renderer/Texture/Texture.h"
+#include "../Renderer/DescriptorHeap.h"
+
+#include "../Misc/Window.h"
+#include "../Misc/AssetLoader.h"
+#include "../Misc/GUI2DElements/Text.h"
+#include "../Misc/GUI2DElements/Font.h"
+
+TextManager::TextManager()
+{
+}
+
+TextManager::~TextManager()
+{
+	for (auto textMap : m_TextMap)
+	{
+		delete textMap.second;
+	}
+	m_TextMap.clear();
+}
+
+std::map<std::string, TextData>* const TextManager::GetTextDataMap()
+{
+	return &m_TextDataMap;
+}
+
+TextData* TextManager::GetTextData(std::string name)
+{
+	return &m_TextDataMap[name];
+}
+
+void TextManager::AddText(std::string name)
+{
+	// Look if a font is chosen, otherwise, set the default font
+	if (m_pFont == nullptr)
+	{
+		m_pFont = AssetLoader::Get()->LoadFontFromFile(L"Arial.fnt");
+	}
+
+	// Default text
+	TextData textData = {};
+	textData.color = { 1.0f, 1.0f, 1.0f, 1.0f };
+	textData.padding = { 0.5f, 0.0f };
+	textData.pos = { 0.5f, 0.5f };
+	textData.scale = { 1.0f, 1.0f };
+	textData.text = L"DEFAULT_TEXT";
+
+	for (auto data : m_TextDataMap)
+	{
+		if (data.first == name)
+		{
+			Log::PrintSeverity(Log::Severity::WARNING, "It already exists a text with the name '%s'! Overwriting text data...\n", name.c_str());
+		}
+	}
+
+	m_TextDataMap[name] = textData;
+}
+
+void TextManager::UploadTextData(std::string name)
+{
+	Renderer::GetInstance().LoadTexture(m_pFont->GetTexture());
+
+	int numOfCharacters = GetNumOfCharacters(name);
+	auto textData = GetTextData(name);
+
+	Renderer* renderer = &Renderer::GetInstance();
+
+	Text* text = new Text(
+		renderer->m_pDevice5,
+		renderer->m_DescriptorHeaps[DESCRIPTOR_HEAP_TYPE::CBV_UAV_SRV],
+		numOfCharacters,
+		m_pFont->GetTexture());
+	text->SetTextData(textData, m_pFont);
+
+	// Look if the text exists
+	bool exists = false;
+	for (auto textMap : m_TextMap)
+	{
+		if (textMap.first == name)
+		{
+			exists = true;
+		}
+	}
+
+	if (exists == true)
+	{
+		replaceText(text, name);
+	}
+	else
+	{
+		submitText(text, name);
+	}
+
+	renderer->submitTextToGPU(text, this);
+
+	renderer->executeCopyOnDemand();
+}
+
+void TextManager::SetFont(Font* font)
+{
+	m_pFont = font;
+}
+
+void TextManager::SetText(std::string text, std::string name)
+{
+	bool exists = false;
+	for (auto data : m_TextDataMap)
+	{
+		if (data.first == name)
+		{
+			m_TextDataMap[name].text = to_wstring(text);
+			exists = true;
+			break;
+		}
+	}
+
+	if (exists == false)
+	{
+		Log::PrintSeverity(Log::Severity::CRITICAL, "The text '%s', does not exist! Could not set text.\n", name.c_str());
+	}
+}
+
+void TextManager::SetPos(float2 textPos, std::string name)
+{
+	bool exists = false;
+	for (auto data : m_TextDataMap)
+	{
+		if (data.first == name)
+		{
+			m_TextDataMap[name].pos = textPos;
+			exists = true;
+			break;
+		}
+	}
+
+	if (exists == false)
+	{
+		Log::PrintSeverity(Log::Severity::CRITICAL, "The text '%s', does not exist! Could not set position.\n", name.c_str());
+	}
```

Comment:
Since it's the third time you loop this way, MAYBE change m_TextDataMap datastructure to something more fitting? Also I would suggest using the std::map.count() or find() to see if the map has an element.